scripts/setup_hooks.py

```diff
@@ -1,31 +1,51 @@
 #!/usr/bin/env python3
 """
-Bootstrap Git pre‑push hook.
+Bootstrap Git pre‑push hook with isolated virtual environment.
 
 ✓ Requires uv to be installed (fails if not available)
-✓ Installs/updates pre‑commit with uv  (global, venv‑proof)
-✓ Registers the repo's pre‑push hook and freezes hook versions
+✓ Creates isolated venv in .git/hooks/linter/.venv/ for hook dependencies
+✓ Installs lintrunner only in the isolated environment
+✓ Creates direct git hook that bypasses pre-commit
 
 Run this from the repo root (inside or outside any project venv):
 
     python scripts/setup_hooks.py
+
+IMPORTANT: The generated git hook references scripts/lintrunner.py. If users checkout
+branches that don't have this file, git push will fail with "No such file or directory".
+Users would need to either:
+1. Re-run the old setup_hooks.py from that branch, or
+2. Manually delete .git/hooks/pre-push to disable hooks temporarily, or
+3. Switch back to a branch with the new scripts/lintrunner.py
 """
 
 from __future__ import annotations
 
+import shlex
 import shutil
 import subprocess
 import sys
 from pathlib import Path
-from typing import Tuple
+
+
+# Add scripts directory to Python path so we can import lintrunner module
+scripts_dir = Path(__file__).parent
+sys.path.insert(0, str(scripts_dir))
+
+# Import shared functions from lintrunner module
+from lintrunner import find_repo_root, get_hook_venv_path
+
+
+# Restore sys.path to avoid affecting other imports
+sys.path.pop(0)
 
 
 # ───────────────────────────────────────────
 # Helper utilities
 # ───────────────────────────────────────────
-def run(cmd: list[str]) -> None:
+def run(cmd: list[str], cwd: Path = None) -> None:
     print(f"$ {' '.join(cmd)}")
-    subprocess.check_call(cmd)
+    subprocess.check_call(cmd, cwd=cwd)
```

Comment:
Why not just forward cmd in escaped list form rather than trying to join into an str yourself

Reply:
Could you please clarify your concern? Over here the list cmd itself is passed through, and the join is only used for logging.  Are you worried about the join for logging, the pass through, or something else?